### DISCUSSION TITLE
From v0.14 to v1 difficulties using Client

### DISCUSSION THREAD

**Discussion by frederikhors** [hyperium/hyper#3462] [General]
Coming from:

```toml
hyper = { version = "0.14.27", default-features = false }
```

I switched to `hyper 1.0.1` but now I'm stuck upgrading the below code:

```rust
use hyper::{client::Client, Body, Request};

pub async fn ping(uri: &str) -> Result<()> {
    let hyper_client = Client::new();

    if let Err(e) = retry(5, 2, || {
        let req = Request::builder()
            .uri(uri.to_string())
            .body(Body::empty())
            .unwrap();

        hyper_client.request(req)
    })
    .await
    {
        return "cannot ping!"
    }

    Ok(())
}
```

Can you help me understand what to use instead of `client::Client` now?

Should I use `hyper_util::client::legacy::Client`?

Why the word `legacy` in the name?

Or should I use all the code in https://github.com/hyperium/hyper/blob/master/examples/client.rs?

**Comment by GlenDC** [comment]
- <https://hyper.rs/guides/1/upgrading/> according to the upgrade guides you can ease your upgrade pain by using `hyper_util::client::legacy::Client` indeed. It has the word legacy in there because it is really more a utility to allow you to continue to use the "old style" client approach
- hyper 1.0 wants to be more low level and thus an actual http client might be too much for that. If you want to not use this legacy utility forever (which isn't recommended anyway I imagine) you might want to either make your own client with hyper's primitives or use something more high level like reqwest;
- I don't think all examples are up to date yet. But might be wrong with that.

Not a maintainer or contributor though, so take what i say with a grain of salt.

**Comment by seanmonstar** [comment]
> Should I use `hyper_util::client::legacy::Client`?

Probably! That's the pooling, connecting client that works just like the previous `hyper::Client`.

> Why the word legacy in the name?

It's the "old" way. It exists to help people upgrade, it works exactly how you're used to. But I think there's a "better" `Client` that can be provided eventually. It just doesn't exist yet.

> Or should I use all the code in the examples

You certainly can. The examples are up-to-date. What you see there are code using hyper's connection-based types. Is that what you want? To manage connections yourself?

If not, you may want to stick with `legacy::Client`, or dare I say it, choose to use [reqwest](https://crates.io/crates/reqwest).
